types/evidence.go

```diff
@@ -342,3 +355,581 @@ func (evl EvidenceList) Has(evidence Evidence) bool {
 	}
 	return false
 }
+
+//-------------------------------------------
+
+// ConflictingHeadersEvidence is primarily used by the light client when it
+// observes two conflicting headers, both having 1/3+ of the voting power of
+// the currently trusted validator set.
+type ConflictingHeadersEvidence struct {
+	H1 *SignedHeader `json:"h_1"`
+	H2 *SignedHeader `json:"h_2"`
+}
+
+var _ Evidence = &ConflictingHeadersEvidence{}
+var _ CompositeEvidence = &ConflictingHeadersEvidence{}
+var _ Evidence = ConflictingHeadersEvidence{}
+var _ CompositeEvidence = ConflictingHeadersEvidence{}
+
+// Split breaks up eviddence into smaller chunks (one per validator except for
+// PotentialAmnesiaEvidence): PhantomValidatorEvidence,
+// LunaticValidatorEvidence, DuplicateVoteEvidence and
+// PotentialAmnesiaEvidence.
+//
+// committedHeader - header at height H1.Height == H2.Height
+// valSet					 - validator set at height H1.Height == H2.Height
+func (ev ConflictingHeadersEvidence) Split(committedHeader *Header, valSet *ValidatorSet) []Evidence {
+	evList := make([]Evidence, 0)
+
+	var alternativeHeader *SignedHeader
+	if bytes.Equal(committedHeader.Hash(), ev.H1.Hash()) {
+		alternativeHeader = ev.H2
+	} else {
+		alternativeHeader = ev.H1
+	}
+
+	// If there are signers(alternativeHeader) that are not part of
+	// validators(committedHeader), they misbehaved as they are signing protocol
+	// messages in heights they are not validators => immediately slashable
+	// (#F4).
+	for i, sig := range alternativeHeader.Commit.Signatures {
+		if sig.Absent() {
+			continue
+		}
+		if !valSet.HasAddress(sig.ValidatorAddress) {
+			evList = append(evList, &PhantomValidatorEvidence{
+				Header: alternativeHeader.Header,
+				Vote:   alternativeHeader.Commit.GetVote(i),
+			})
+		}
+	}
+
+	// If ValidatorsHash, NextValidatorsHash, ConsensusHash, AppHash, and
+	// LastResultsHash in alternativeHeader are different (incorrect application
+	// state transition), then it is a lunatic misbehavior => immediately
+	// slashable (#F5).
+	var invalidField string
+	switch {
+	case !bytes.Equal(committedHeader.ValidatorsHash, alternativeHeader.ValidatorsHash):
+		invalidField = "ValidatorsHash"
+	case !bytes.Equal(committedHeader.NextValidatorsHash, alternativeHeader.NextValidatorsHash):
+		invalidField = "NextValidatorsHash"
+	case !bytes.Equal(committedHeader.ConsensusHash, alternativeHeader.ConsensusHash):
+		invalidField = "ConsensusHash"
+	case !bytes.Equal(committedHeader.AppHash, alternativeHeader.AppHash):
+		invalidField = "AppHash"
+	case !bytes.Equal(committedHeader.LastResultsHash, alternativeHeader.LastResultsHash):
+		invalidField = "LastResultsHash"
```

Comment:
Is it possible to have multiple invalid fields - is it okay if we only submit one

Reply:
It is possible

Reply:
thought: we can make it an array

Reply:
Actually I think one is sufficient